# Test Blocks

<EpicVideo url="https://www.epicweb.dev/workshops/testing-fundamentals/test-structure/isolating-tests-with-test-blocks/solution" />

I start with defining the `title` and `callback` arguments on the `test()` function. Those will represent the _test's title_ and the _test itself_, respectively. Next, I invoke the `callback()` function, which will run whichever test we provide. Since failed assertions throw errors, I wrap the `callback()` in a `try/catch` block to prevent the process from exiting on failed assertions and also to print those nicely in the terminal's output.

```ts filename=greet.ts highlight=2,5 nonumber
function test(title: string, callback: () => void) {
	try {
		callback()
		console.log(`✓ ${title}`)
	} catch (error) {
		console.error(`✗ ${title}`)
		console.error(error, '\n')
	}
}
```

Then, I wrap our existing tests in the `test()` function, giving it a meaningful title so the expectation behind each test is clear.

```ts filename=greet.ts highlight=1,5 nonumber
test('returns a greeting message for the given name', () => {
	expect(greet('John')).toBe('Hello, John!')
})

test('returns a congratulation message for the given name', () => {
	expect(congratulate('Sarah')).toBe('Congrats, Sarah!')
})
```

Now, whenever a test fails, we can immediately see its title and the relevant assertion error below.

```sh nonumber
npx tsx greet.ts
```

```txt nocopy lines=1-3
✗ returns a greeting message for the given name
Error: Expected Hello, ${name}! to equal to Hello, John!
    at /Users/kody/testing-fundamentals/greet.ts:10:24

✓ returns a congratulation message for the given name
```
